### PR TITLE
Build standalone Python wheels

### PR DESCRIPTION
**Pull request by vmarkovtsev**
I call "a standalone wheel" a package which does not require the libartm installation - aka "tensorflow style". Those wheels can be safely pushed on PyPi and will just work.

I had to change:

* `python/CMakeLists.txt`: add another custom target to execute ` setup.py bdist_wheel`
* `python/setup.py`: hack setuptools to include the needed shared library
* `python/artm/wrapper/api.py`: refactor `_load_cdll()` to load the library from several places - the common practice. The env variable is the most important, then the default relative path and finally the absolute path in the package root.
* docs - as far as I could

Besides, there seems to be a bug in TravisCI configuration regarding python command detection. There is no explicit `-DPYTHON` in the build script, and by default cmake chooses `python2` which prevents from testing py3.x:

```
if (MSVC OR APPLE)
set(PYTHON python CACHE INTERNAL "Python command")
else (MSVC OR APPLE)
set(PYTHON python2 CACHE INTERNAL "Python command")
endif (MSVC OR APPLE)
```

`bdist_wheel` command helped to find this bug: there is no `wheel` package installed in py2 env in Travis.
I dared to remove `python2` branch completely.

### CODE REVIEW

**Comment by JeanPaulShapo** [comment]
Great job! I'll try to review this PR in the weekend.

**Comment by vmarkovtsev** [comment]
Ping.

**Comment by JeanPaulShapo** [comment]
SYN-ACK

**Comment by JeanPaulShapo** [comment]
Could you please describe common usage pattern of this method? Should it be used **only** by skilled developers just to upload to PyPi? Also, if you have some links to useful material, it'll be great if you share it with us.

**Comment by vmarkovtsev** [comment]
The usage pattern is pretty much the same as with Tensorflow: we install the wheel inside our docker environment without the need to compile the library which is so painful in that environment. It is actually easier and more reliable than installing the whole bigartm suite with CLI and `protoc` (btw, is it really needed?), and is guaranteed to work in Windows/macOS with possibly screwed library paths.

I would suggest to upload this wheel to PyPi - make a binary release after applying `auditwheel repair`. There will be no need for end users to build the package from source.

File: python/artm/wrapper/api.py [hunk 1]
```diff
@@ -42,22 +42,35 @@ def _load_cdll(self, lib_name):
...
+        
+        if lib_name is not None:
+            lib_names.append(lib_name)
+            
+        env_lib_name = os.environ.get('ARTM_SHARED_LIBRARY')
+        if env_lib_name is not None:
+            lib_names.append(env_lib_name)
+        
+        lib_names.append(default_lib_name)
+        lib_names.append(os.path.join(os.path.dirname(__file__), "..", default_lib_name))
```
Comment (JeanPaulShapo) [inline python/artm/wrapper/api.py]: Maybe there is have to be one additional '..' folder?

Comment (JeanPaulShapo) [inline python/artm/wrapper/api.py]: Thanks, I got it.

**Comment by JeanPaulShapo** [comment]
Generally this PR seems rather OK, however there are some few questions left:

* Should we check installation from wheels on continuous integration?
* What credentials (e.g. maintainers and email) to supply when uploading wheel to PyPI? There is an example in [the file from outdated PR](https://github.com/JeanPaulShapo/bigartm/blob/cf6773f15aecf378dea4856f156509e3b9dfb15d/setup.py). 

And sorry, I really need some time to refresh my mind on this `setuptools`-`distutils` topic.

**Comment by JeanPaulShapo** [comment]
And you'd better rebase the branch.

**Comment by vmarkovtsev** [comment]
* Checking the installation: yes, we should, but let's do it in a separate PR
* You are the maintainers :) Take for example [ours](https://github.com/src-d/ast2vec/blob/master/setup.py) I am not uploading anything myself anyway.

Refreshing the topic: Armin's nice blog post http://lucumr.pocoo.org/2012/6/22/hate-hate-hate-everywhere/

**Comment by vmarkovtsev** [comment]
...and rebased.

**Comment by JeanPaulShapo** [comment]
Thanks for the article. I'll try to respond in two days.

**Comment by JeanPaulShapo** [comment]
@vmarkovtsev 

As I've got some information about BigARTM Windows distribution from @ofrei, you need not only `artm.dll` library but also Visual Studio runtime libraries (namely, `msvcp140.dll` and `vcruntime140.dll`). Therefore, in case of Windows we mustn't consider path to python package root - you need to add this path only when `'win' no in sys.platform`.
Now, consider we add check for windows/non-windows. Then, if you intend to use this code only as some kind of local Python binary package on non-windows, this PR seem OK for me. I think we can have some separate PR to fully integrate our ibrary with PyPI.

So, if I understand you correctly, you only have to rebase PR again before merging.

File: python/artm/wrapper/api.py (hunk 1 above)
Comment (JeanPaulShapo) [inline python/artm/wrapper/api.py]: I think we mustn't use for Windows, i.e. include this path only conditionallt.

**Comment by ofrei** [comment]
@vmarkovtsev @JeanPaulShapo Sorry for my delay. Excellent PR, thanks! I've tested that this PR doesn't break current installation steps (see [these](https://github.com/bigartm/bigartm-install/pull/5) CI runs in ``bigartm-install`` repo). Since current scenarios work we can safely merge this PR, and address the remaining questions separately (including Windows story for the bigartm wheel, putting it on PyPi, etc).

**Comment by vmarkovtsev** [comment]
Yeah, sorry guys, got a bit sick lately, should have rebased this myself.

**Comment by ofrei** [comment]
@vmarkovtsev no problem! if it takes 2 weeks for us to review it feels to me like our responsibility to bring in this change

**Comment by JeanPaulShapo** [comment]
@vmarkovtsev Could you please add some comments about my thoughts on Windows distribution?

**Comment by ofrei** [comment]
@JeanPaulShapo Regarding Windows wheel... when we produce binary package for Windows there is no problem with VS runtime libraries --- thanks to CMake ``InstallRequiredSystemLibraries`` flag. I'm not sure how to automatically include the right version of the VS runtime libraries into python wheel, because it depends on the version of visual studio that is used. I think it needs to be tested separately. For now it is perfect if we can make Wheel work on Linux.

**Comment by JeanPaulShapo** [comment]
@ofrei OK, in that case there is no problem left. Feel free for merge corresponding PR after CI.

**Comment by ofrei** [comment]
Merged through #838

**Pull request closed without merging**